hudi-utilities/src/main/java/org/apache/hudi/utilities/HoodieMetadataTableValidator.java

```diff
@@ -0,0 +1,337 @@
+    if (!latestFileSlicesFromMetadataTable.equals(latestFileSlicesFromFS)) {
+      String message = "Validation of metadata get latest file slices for partition " + partitionPath + " failed."
+          + "Latest file list from metadata: " + latestFileSlicesFromMetadataTable
+          + "Latest file list from direct listing: " + latestFileSlicesFromFS;
+      LOG.error(message);
+      throw new RuntimeException(message);
+    } else {
+      LOG.info("Validation of getLatestFileSlices success.");
+    }
+
+    List<HoodieBaseFile> latestFilesFromMetadata = metaFsView.getLatestBaseFiles(partitionPath).collect(Collectors.toList());
+    List<HoodieBaseFile> latestFilesFromFS = fsView.getLatestBaseFiles(partitionPath).collect(Collectors.toList());
+
+    LOG.info("Latest base file from metadata: " + latestFilesFromMetadata);
+    LOG.info("Latest base file from direct listing: " + latestFilesFromFS);
+    if (!latestFilesFromMetadata.equals(latestFilesFromFS)) {
+      String message = "Validation of metadata get latest base file for partition " + partitionPath + " failed. "
+          + "Latest base file from metadata: " + latestFilesFromMetadata
+          + "Latest base file from direct listing: " + latestFilesFromFS;
+      LOG.error(message);
+      throw new RuntimeException(message);
+    } else {
+      LOG.info("Validation of getLatestBaseFiles success.");
+    }
+  }
+
+  private HoodieTableFileSystemView createHoodieTableFileSystemView(HoodieSparkEngineContext engineContext, boolean enableMetadataTable) {
+
+    HoodieMetadataConfig metadataConfig = HoodieMetadataConfig.newBuilder()
+        .enable(enableMetadataTable)
+        .withAssumeDatePartitioning(false)
+        .build();
+
+    return FileSystemViewManager.createInMemoryFileSystemView(engineContext,
+        metaClient, metadataConfig);
+  }
+
+  protected boolean compareCollectionResult(Collection<String> c1, Collection<String> c2) {
+    return c1.size() == c2.size() && c1.containsAll(c2) && c1.containsAll(c2);
+  }
+
+  public class AsyncMetadataTableValidateService extends HoodieAsyncService {
+    private final transient ExecutorService executor = Executors.newSingleThreadExecutor();
+
+    @Override
+    protected Pair<CompletableFuture, ExecutorService> startService() {
+      return Pair.of(CompletableFuture.supplyAsync(() -> {
+        while (true) {
+          try {
+            long start = System.currentTimeMillis();
+            doMetaTableValidation();
+            long toSleepMs = cfg.minValidateIntervalSeconds * 1000 - (System.currentTimeMillis() - start);
+
+            if (toSleepMs > 0) {
+              LOG.info("Last validate ran less than min validate interval: " + cfg.minValidateIntervalSeconds + " s, sleep: "
+                  + toSleepMs + " ms.");
+              Thread.sleep(toSleepMs);
+            }
+          } catch (Exception e) {
+            LOG.error("Shutting down AsyncMetadataTableValidateService due to exception", e);
```

Comment:
should we explicitly check for validation failures and ignore only on such cases. What incase there was some other exception and we keep running the job in a continuous mode. 
so, probably when you throw RuntimeException incase of validation mismatch, you can set some explicit cause or msg and add add guards here based on that.

Reply:
Nice catch. Throw `HoodieValidationException` when Validate failed.
Thanks a lot for your review.